rmd_files/strings.Rmd

```diff
@@ -3,565 +3,268 @@
-#> [1] "The"   "Glue"  "It"    "These" "Rice"  "The"
+We can revisit the example from earlier, where we wanted to identify strings containing the colour 'red':
 
+```{r, purl=purl_example_code}
+colours <- c("scarlet red", "ultramarine blue", "cadmium red", "cobalt blue", "weathered")
+stringr::str_detect(colours, "\\bred\\b")
 ```
 
+The word boundary regex allows us to exclude words like "weathered" when looking for the word "red".
 
-However, the third sentence begins with “It’s”. To catch this, I’ll change the regular expression to require the string to begin with a letter, but allow for a subsequent apostrophe.
 
+## Extracting a matched pattern
 
-```{r, include=show_solution, purl=purl_solutions}
-stringr::str_extract(sentences, "[A-Za-z][A-Za-z']*") %>% head()
-#> [1] "The"   "Glue"  "It's"  "These" "Rice"  "The"
+We can use the `str_extract()` function to extract strings that match a particular pattern:
 
+```{r, purl=purl_example_code}
+colours <- c("scarlet red", "ultramarine blue", "cadmium red", "cobalt blue", "cerulean blue")
+stringr::str_extract(colours, "blue")
 ```
 
----
 
-Part 2: This pattern finds all words ending in 'ing'.
+## Replacing a matched pattern
 
-```{r, include=show_solution, purl=purl_solutions}
-pattern <- "\\b[A-Za-z]+ing\\b"
-sentences_with_ing <- stringr::str_detect(sentences, pattern)
-unique(unlist(stringr::str_extract_all(sentences[sentences_with_ing], pattern))) %>%
-  head()
-#> [1] "spring"  "evening" "morning" "winding" "living"  "king"
+You can use the `str_replace()` and `str_replace_all()` functions to find and replace parts of a string. `str_replace()` replaces the first instance of the pattern, whereas `str_replace_all()` replaces all instances of the pattern. Here's an alternative version of an example we saw earlier, using a different approach to replace "value" with "item":
 
+```{r, purl=purl_example_code}
+x <- c("First value", "Second value", "Third value")
+# Replace 'value' with 'item'
+stringr::str_replace(x, "value", "item")
 ```
 
 ---
 
-Part 3: Finding all plurals cannot be correctly accomplished with regular expressions alone. Finding plural words would at least require morphological information about words in the language. See WordNet for a resource that would do that. However, identifying words that end in an “s” and with more than three characters, in order to remove “as”, “is”, “gas”, etc., is a reasonable heuristic.
+Regular expressions are also useful for string replacement. Here's a example that replaces characters that aren't letters or numbers with an underscore:
 
-```{r, include=show_solution, purl=purl_solutions}
-
-unique(unlist(stringr::str_extract_all(sentences, "\\b[A-Za-z]{3,}s\\b"))) %>%
-  head()
-#> [1] "planks" "days"   "bowls"  "lemons" "makes"  "hogs"
+```{r, purl=purl_example_code}
+colours <- c("scarlet...red", "ultramarine.blue", "cadmium_red", "cobalt blue", "cerulean-blue")
 
+# Replace the first character that isn't a letter or number with an underscore
```

Comment:
This returns

```
[1] "scarlet_..red"    "ultramarine_blue" "cadmium_red"      "cobalt_blue"     
[5] "cerulean_blue"   
[1] "scarlet___red"    "ultramarine_blue" "cadmium_red"      "cobalt_blue"     
[5] "cerulean_blue"  
```

Actually my regex isn't good enough for me to be able to suggest the solution, so I leave it to you!

Reply:
As discussed - this was just to demonstrate the difference in behaviour between `str_replace` and `str_replace_all`, although could possibly think of some better example data...

I've added an extra sentence underneath to point out what's happened.